fish-rust/src/builtins/string.rs

```diff
@@ -0,0 +1,518 @@
+
+                streams.err.append(L!("string "));
+                streams.err.append(error_msg);
+            }
+            NotANumber => {
+                string_error!(streams, BUILTIN_ERR_NOT_NUMBER, cmd, optarg.unwrap());
+            }
+            UnknownOption => {
+                string_unknown_option(parser, streams, cmd, args[optind - 1]);
+            }
+        }
+    }
+
+    fn retval(&self) -> Option<c_int> {
+        STATUS_INVALID_ARGS
+    }
+}
+
+#[derive(Default, PartialEq, Clone, Copy)]
+enum Direction {
+    #[default]
+    Left,
+    Right,
+}
+
+pub(self) fn width_without_escapes(ins: &wstr, start_pos: usize) -> usize {
+    let mut width: i32 = 0;
+    for c in ins[start_pos..].chars() {
+        let w = fish_wcwidth_visible(c);
+        // We assume that this string is on its own line,
+        // in which case a backslash can't bring us below 0.
+        if w > 0 || width > 0 {
+            width += w;
+        }
+    }
+    // ANSI escape sequences like \e\[31m contain printable characters. Subtract their width
+    // because they are not rendered.
+    let mut pos = start_pos;
+    while let Some(ec_pos) = ins.slice_from(pos).find_char('\x1B') {
+        pos += ec_pos;
+        if let Some(len) = escape_code_length(ins.slice_from(pos)) {
+            let sub = &ins[pos..pos + len];
+            for c in sub.chars() {
+                width -= fish_wcwidth_visible(c);
+            }
+            // Move us forward behind the escape code,
+            // it might include a second escape!
+            // E.g. SGR0 ("reset") is \e\(B\e\[m in xterm.
+            pos += len - 1;
+        } else {
+            pos += 1;
+        }
+    }
+    // we subtracted less than we added
+    debug_assert!(width >= 0, "line has negative width");
+    return width as usize;
+}
+
+pub(self) fn escape_code_length(code: &wstr) -> Option<usize> {
+    use crate::ffi::escape_code_length_ffi;
+    use crate::wchar_ffi::wstr_to_u32string;
+
+    match escape_code_length_ffi(wstr_to_u32string(code).as_ptr()).into() {
+        -1 => None,
+        n => Some(n as usize),
+    }
+}
+
+/// A helper type for extracting arguments from either argv or stdin.
+pub(self) struct Arguments<'args, 'iter> {
+    /// The list of arguments passed to the string builtin.
+    args: &'iter [&'args wstr],
+    /// If using argv, index of the next argument to return.
+    argidx: &'iter mut usize,
+    /// If set, when reading from a stream, split on newlines.
+    split_on_newline: bool,
+    /// Buffer to store what we read with the BufReader
+    /// Is only here to avoid allocating every time
+    buffer: Vec<u8>,
+    /// If not using argv, we read with a buffer
+    reader: Option<BufReader<File>>,
+}
+
+impl Drop for Arguments<'_, '_> {
+    fn drop(&mut self) {
+        if let Some(r) = self.reader.take() {
+            // we should not close stdin
+            std::mem::forget(r.into_inner());
```

Comment:
As far as I know `File` is just a wrapper around a `OwnedFd`, which does not contain any allocations. This is the prime example of [`std::mem::forget`](https://doc.rust-lang.org/std/mem/fn.forget.html), so if it has the potential to cause a memory leak I would find that a large deficiency in Rust's docs. Note that unlike `Box::leak`, references to the memory might become invalid. This code is also run in ASAN, so that would already have detected a memory leak?

Semi-relevant: `Option<File>` has the same representation as `File`, since `-1` is not a valid Fd

Reply:
Ok, if this is idiomatic then I'm fine with it as written.